#42: ThreadWorker

MHajoha opened the pull request:
Implementiert einen `ThreadWorker`, der das Paket in einem Thread im Hauptprozess ausführt, wo das lokale debuggen von Paketcode etwas einfacher ist. Der Großteil dieses PRs besteht allerdings aus dem Refactoring, das dafür nötig ist. Dabei werden die alten `WorkerProcessBase` und `WorkerProcess` auseinandergezogen in:
* `Worker` - Öffentliches Interface für Worker, nur abstrakte Methoden. Der Übersichtlichkeit halber ist das von `BaseWorker` getrennt
* `BaseWorker` - Teile der `Worker`-Impl, die für `SubprocessWorker` und `ThreadWorker` gleichermaßen relevant sind. (Und vermutlich auch für zukünftige andere Implementierungen)
* `SubprocessWorker` - Was auf der Dose steht. Funktional sollte das nicht anders sein als der alte `WorkerProcess`. Für die Übersichtlichkeit habe ich auch das Buffering in eine andere Klasse gezogen, aber das ist sicherlich Geschmackssache.

Neu ist der `ThreadWorker`. Der führt die Runtime in einem eigenen Thread aus und kommuniziert damit über zwei Pipes (eine für hin, eine für zurück). Um nicht überall mit 4 FDs jonglieren zu müssen, habe ich eine Klasse DuplexPipe geschrieben, die sich darum kümmert. 

Damit die beiden Connection-Klassen damit auch klar kommen, habe ich die so verändert, dass sie mit allen Objekten arbeiten können, die `SupportsWrite` (statt `StreamWriter`/`RawIOBase`) bzw. `SupportsRead` (statt `BufferedReader`) oder `SupportsAsyncRead` (statt `StreamReader`) unterstützen.

MHajoha committed:
feat: add ThreadWorker for local testing

MHajoha committed:
test: add tests for workers

MHajoha committed:
refactor: raise WorkerStartError when worker dies during init

Was WorkerNotRunningError, which is a bit misleading since initialization happens in start().

MartinGauk commented on questionpy_server/api/routes.py:
```diff
@@ -57,8 +58,8 @@ async def post_options(request: web.Request, package: Package, question_state: O
     async with qpyserver.worker_pool.get_worker(package_path, 0, data.context) as worker:
-        options = await worker.get_options_form_definition()
-    return json_response(data=options)
+        response = await worker.send_and_wait_response(GetOptionsFormDefinition(), GetOptionsFormDefinition.Response)
```
Mir gefällt es nicht so, dass man sich mit den Messages befassen muss. Besonders für das SDK sind das unnötige Details und dann ist es besser, wenn es ein klares Interface gibt, um auf ein Paket/Worker zugreifen zu können.

MHajoha replied:
Ich habe die Worker-Methoden get_manifest, get_options_form_definition und create_question_from_options wieder hinzugefügt

MartinGauk commented on questionpy_server/utils/streams.py:
```diff
@@ -0,0 +1,96 @@
+            size = len(data)
+            written = 0
+            while written < size:
+                written += self._transmit.write(data[written:])
```
Du könntest hier `memoryview` benutzen, damit die Daten nicht kopiert werden müssen.

Bei `read` könnte `memoryview` ebenfalls einen Kopiervorgang einsparen, falls `readinto` bei Pipes verfügbar ist.

MHajoha replied:
Sorgt `data[written:]` im Ernst dafür dass die Daten kopiert werden? :laughing: Ich hatte naiv erwartet, dass im Hintergrund einfach `&data + written` passiert. Aber okay, also `memoryview`.

MHajoha replied:
Wegen des reads: Ich hatte mich auf die Type Hints in IDEA verlassen, nach denen die streams `typing.BytesIO` sind. Anscheinend kann ich mich aber darauf verlassen, dass sie `(BufferedReader, FileIO)` sind. 

`BufferedReader.read` macht schon das, was ich da gebaut habe (blockt also bis `size` bytes da sind oder EOF da ist), `read` wird also ein one-liner.

Für `write` passt Buffering natürlich nicht. Da könnte ich höchstens `write() + flush()` benutzen, finde die eigene Implementierung da aber besser verständlich.

MartinGauk commented on questionpy_server/worker/connection.py:
```diff
@@ -0,0 +1,49 @@
+    InvalidMessageIdError
+
+
+class ServerToWorkerConnection(AsyncIterator[MessageToServer]):
```
Pylint meckert über die Code-Ähnlichkeit von `ServerToWorkerConnection` und `WorkerToServerConnection`. Ich denke, dass wir das schon so belassen können, aber wir sollten die Meldung unterdrücken.

MHajoha replied:
Ich wüsste auch nicht, wie man die beiden vereinen könnte, wo sie ja eigentlich das gleiche machen. [Nur die Farbe ist anders](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/). Habe die Warnung ausgemacht

MartinGauk commented on questionpy_server/worker/pool.py:
```diff
@@ -56,14 +35,15 @@ def memory_available(self, size: int) -> bool:
-        :param context: context id within the lms
+        :param _lms: id of the LMS
+        :param _context: context id within the lms
+        :param worker_type: worker implementation
```
Diese Zeile müsste oben stehen.

MHajoha committed:
refactor: re-add request methods to Worker

get_manifest, get_options_form_definition, and create_question_from_options. Also make send_and_wait_response private since it isn't used externally.

MHajoha committed:
refactor: avoid copy in DuplexPipe.write, simplify read

MHajoha committed:
style: silence pylint connection duplicate warning
